apps/emqx_prometheus/src/emqx_prometheus.erl

```diff
@@ -672,15 +673,53 @@ do_cluster_data(Labels) ->
 %%========================================
 
 emqx_metric_data(MetricNameTypeKeyL, Mode) ->
+    emqx_metric_data(MetricNameTypeKeyL, Mode, _Acc = #{}).
+
+emqx_metric_data(MetricNameTypeKeyL, Mode, Acc) ->
     Metrics = emqx_metrics:all(),
     lists:foldl(
-        fun({Name, _Type, MetricKAtom}, AccIn) ->
-            AccIn#{Name => [{with_node_label(Mode, []), ?C(MetricKAtom, Metrics)}]}
+        fun
+            ({_Name, _Type, undefined}, AccIn) ->
+                AccIn;
+            ({Name, _Type, MetricKAtom}, AccIn) ->
+                AccIn#{Name => [{with_node_label(Mode, []), ?C(MetricKAtom, Metrics)}]}
         end,
-        #{},
+        Acc,
         MetricNameTypeKeyL
     ).
 
+client_metric_data(Mode) ->
+    Acc = listener_shutdown_counts(Mode),
+    emqx_metric_data(client_metric_meta(), Mode, Acc).
+
+listener_shutdown_counts(Mode) ->
+    Listeners = emqx_listeners:list(),
+    ListenerIds = lists:map(fun({Id, #{bind := Bind}}) -> {Id, Bind} end, Listeners),
+    Counts =
+        lists:flatmap(
+            fun({Id, Bind}) ->
+                case emqx_listeners:shutdown_count(Id, Bind) of
+                    {error, _} -> [];
+                    Count -> Count
+                end
+            end,
+            ListenerIds
+        ),
+    CountsByReason =
+        lists:foldl(
+            fun({Reason, Count}, Acc) ->
+                maps:update_with(Reason, fun(C) -> C + Count end, Count, Acc)
+            end,
+            #{},
+            Counts
+        ),
```

Comment:
Naturally wonder: why aggregate manually across listeners? In general, aggregation is a prometheus' job after all. I feel that having counts per listener and per reason is slightly more useful, not to mention is simpler to emit.

Reply:
Fair, the listener type and name could just be more labels.